core/hashtree/gc.go

```diff
@@ -0,0 +1,232 @@
+			g.keysRemoved += uint64(len(g.delkeys))
+		}
+		g.writeLock.Unlock()
+		var r util.Range
+		if nextKey == nil {
+			r = *util.BytesPrefix(g.prefix)
+		} else {
+			r.Limit = nextKey
+		}
+		r.Start = startKey
+		g.db.LDB().CompactRange(r)
+	}()
+
+	g.gcBlockHasData = g.hasData(g.gcBlock)
+	it.Seek(startKey)
+	for it.Valid() {
+		key := common.CopyBytes(it.Key())
+		//log.Info("key", "key", key)
+		if len(key) < len(g.prefix) || !bytes.Equal(key[:len(g.prefix)], g.prefix) {
+			return nil
+		}
+		if maxEntries == 0 {
+			nextKey = key
+			return nextKey
+		}
+
+		if len(key) >= len(g.prefix)+33 && key[len(key)-1] == 0 {
+			it.Next()
+			var refkeys [][]byte
+			for it.Valid() {
+				refkey := common.CopyBytes(it.Key())
+				//log.Info("ref", "key", refkey)
+				if len(refkey) >= len(key) && bytes.Equal(refkey[:len(key)-1], key[:len(key)-1]) {
+					if len(refkey) == len(key)+8 && refkey[len(key)+7] == 1 {
+						refkeys = append(refkeys, refkey)
+					} else {
+						log.Error("Invalid hashtree ref", "key", refkey)
+					}
+					it.Next()
+				} else {
+					break
+				}
+			}
+			g.gcEntry(key, refkeys)
+			maxEntries--
+		} else {
+			log.Error("Invalid hashtree entry", "key", key)
+			it.Next()
+		}
+	}
+	return nil
+}
+
+func (g *GarbageCollector) gcEntry(key []byte, refkeys [][]byte) {
+	refcount := len(refkeys)
+	keylen := len(key)
+	oldrefs := 0
+	for oldrefs < refcount {
+		version := binary.BigEndian.Uint64(refkeys[oldrefs][keylen-1 : keylen+7])
+		if version >= g.gcBlock {
+			break
+		}
+		oldrefs++
+	}
+
+	removerefs := 0
+	if oldrefs > 0 {
+		removerefs = oldrefs - 1
+		if oldrefs == refcount && !g.gcBlockHasData(key[len(g.prefix):keylen-33], key[keylen-33:keylen-1]) {
+			removerefs = refcount
+		}
+	}
+
+	g.keysChecked++
+	if removerefs == refcount {
+		g.delkeys = append(g.delkeys, key)
+	}
+	g.refsChecked += uint64(refcount)
+	g.refsRemoved += uint64(removerefs)
+	for i := 0; i < removerefs; i++ {
+		g.db.Delete(refkeys[i])
+	}
+}
+
+func (g *GarbageCollector) FullGC(block uint64) {
+	log.Info("Starting full GC", "block", block)
+	g.gcBlock = block
+	key := g.prefix
+	for key != nil {
+		key = g.run(key, 10000)
+		k := key
+		if len(k) > 8 {
+			k = k[:8]
+		}
+		log.Info("Running...", "key", k, "keys checked", g.keysChecked, "keys removed", g.keysRemoved, "refs checked", g.refsChecked, "refs removed", g.refsRemoved)
+	}
+	log.Info("Finished full GC", "keys checked", g.keysChecked, "keys removed", g.keysRemoved, "refs checked", g.refsChecked, "refs removed", g.refsRemoved)
+}
+
+func (g *GarbageCollector) BackgroundGC(currentBlock func() *types.Block, processing, stop *int32, wg *sync.WaitGroup) {
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+
+		var gcCounter uint64
+		key := g.prefix
+
+		for atomic.LoadInt32(stop) == 0 {
+			wc := atomic.LoadUint64(&g.writeCounter)
+			diff := wc - gcCounter
+			if diff > 10000 {
+				gcCounter = wc - 10000
+				diff = 10000
+			}
+			if diff >= 100 && atomic.LoadInt32(processing) == 0 {
```

Comment:
I don't understand.. You verify that `processing` is not happening here, but it there's nothing preventing it from being `1` immediately after this if-statement has been evaluated. Is that ok?

Reply:
It is. Processing can start while doing GC, it is usually prevented in order to not hinder block processing. If new state trie nodes are written while collecting keys to be deleted (between the two instances of writeLock being locked in GC) then the collected node entries are not deleted in order to avoid collision. The references still are, but those are concurrency-safe.